server/app/controllers/applicant/UpsellController.java

```diff
@@ -165,4 +177,56 @@ public CompletionStage<Result> considerRegister(
               throw new RuntimeException(ex);
             });
   }
+
+  /** Download a PDF file of the application to the program. */
+  @Secure
+  public CompletionStage<Result> download(
+      Http.Request request, long programId, long applicationId, long applicantId)
+      throws ProgramNotFoundException {
+    CompletableFuture<ProgramDefinition> program =
+        programService.getProgramDefinitionAsync(programId).toCompletableFuture();
+    Optional<CiviFormProfile> profileMaybe = profileUtils.currentUserProfile(request);
+    CiviFormProfile profile =
+        profileMaybe.orElseThrow(
+            () -> new NoSuchElementException("User authorized as applicant but no profile found"));
+
+    CompletableFuture<ApplicantPersonalInfo> applicantPersonalInfo =
+        applicantService.getPersonalInfo(applicantId).toCompletableFuture();
+
+    CompletableFuture<Account> account =
+        applicantPersonalInfo
+            .thenComposeAsync(
+                v -> checkApplicantAuthorization(request, applicantId), httpContext.current())
+            .thenComposeAsync(v -> profile.getAccount(), httpContext.current())
+            .toCompletableFuture();
+
+    CompletableFuture<Optional<Application>> applicationMaybe =
+        applicationService.getApplicationAsync(applicationId, program).toCompletableFuture();
+
+    return CompletableFuture.allOf(applicationMaybe, program, account)
+        .thenApplyAsync(
+            check -> {
+              PdfExporter.InMemoryPdf pdf =
+                  pdfExporterService.generatePdf(applicationMaybe.join().get());
```

Comment:
**suggestion (blocking):** There is a security vulnerability here where someone signed in as an applicant could make a request to this endpoint with an `applicationId` for an application that isn't theirs. Add a check that ensures the account requesting the application has permission to read it.

Reply:
I believe the checkApplicantAuthorization(applicantId) checks the same. @zarifzm  - Please confirm